File: examples/Cpp/ICP_SLAM.py
import os
import sys
import numpy as np
import open3d
import time
import math


def visualize_callback(v):
    timera = time.time()
    global i, transform, trajectory, orientation
    try:
        pc = open3d.read_point_cloud(data_dir_path + file_list[i])
    except IndexError:
        return False
    if len(pc_map.points) < 500:
        pc_map.points = pc.points
        return True

    pc = open3d.voxel_down_sample(pc, voxel_size)
    pc.transform(transform)
    open3d.estimate_normals(pc, search_param=search_param)

    info = open3d.registration_icp(
        source=pc,
        target=pc_map,
        max_correspondence_distance=threshold,
        estimation_method=open3d.TransformationEstimationPointToPlane(),
        criteria=open3d.ICPConvergenceCriteria())

    transform = info.transformation.dot(transform)

    trajectory.points = open3d.Vector3dVector(
        np.vstack([np.asarray(trajectory.points),
                   np.array(transform[:3, 3])]))
    pc.transform(info.transformation)
    mesh.transform(info.transformation)
    pc_map.points = open3d.Vector3dVector(
        np.vstack([np.asarray(pc_map.points)[:],
                   np.asarray(pc.points)]))
    pc_map.points = open3d.voxel_down_sample(pc_map, voxel_size).points

    print("Now is the ", i, " Scan.  ", "The size of map is ",
          len(pc_map.points))

    open3d.estimate_normals(pc_map, search_param=search_param)
    i += 1
    print("Time is ", time.time() - timera)
    return True


if __name__ == '__main__':
    data_dir_path = 'point_cloud/'

    file_list = os.listdir(data_dir_path)
    file_list = sorted(file_list)
    print(file_list)
    print(len(file_list))

    pc_map = open3d.read_point_cloud(data_dir_path + file_list[0])
    i = 1

    voxel_size = 0.5
    threshold = 0.5
    search_param = open3d.KDTreeSearchParamHybrid(
        radius=voxel_size * 2, max_nn=30)
    transform = np.eye(4)

    trajectory = open3d.PointCloud()

    mesh = open3d.create_mesh_coordinate_frame(size=5)
    open3d.draw_geometries_with_animation_callback([pc_map, mesh, trajectory],
                                                   visualize_callback,
                                                   width=1000,
                                                   height=600)

File: examples/Cpp/ICP_SLAM.cpp
#include <iostream>
#include <memory>
#include <thread>
#include <sys/types.h>
#include <dirent.h>

#include "Core/Core.h"
#include "IO/IO.h"
#include "Visualization/Visualization.h"

double vox_s = 0.50;
double max_d = 0.50;
int i = 1;

std::vector<std::string> get_file_names(const std::string& dir_name)
{
    std::vector<std::string> file_names;
    DIR* dirp = opendir(dir_name.c_str());
    struct dirent * dp;
    while ((dp = readdir(dirp)) != NULL) {
        file_names.push_back(dp->d_name);
    }
    closedir(dirp);
    return file_names;
}


int main() {
    std::string dir_name = "point_cloud";
    std::vector<std::string> recordfilename = get_file_names(dir_name);
    printf("There are %d files.\n", recordfilename.size());

    auto target = std::make_shared<open3d::PointCloud>();
    auto target_down = std::make_shared<open3d::PointCloud>();

    auto mesh = open3d::CreateMeshCoordinateFrame(3.0);

    auto Search_para = open3d::KDTreeSearchParamHybrid(vox_s * 2, 30);
    open3d::SetVerbosityLevel(open3d::VerbosityLevel::VerboseAlways);
    open3d::ReadPointCloud(recordfilename[0], *target);
    target_down = open3d::VoxelDownSample(*target, vox_s);
    open3d::EstimateNormals(*target_down, Search_para);
    Eigen::Matrix4d current_transformation = Eigen::Matrix4d::Identity();
    open3d::DrawGeometriesWithAnimationCallback(
        {target_down, mesh},
        [&](open3d::Visualizer *vis) {
            int time = clock();
            auto source_down = std::make_shared<open3d::PointCloud>();

            if (open3d::ReadPointCloud(recordfilename[i], *source_down)) {
                int timetime = clock();
                printf("Time for reading file = %dms\n", clock() - time);
                time = clock();
                source_down->points_ =
                    open3d::VoxelDownSample(*source_down, vox_s)->points_;
                printf("Time for down sampling = %dms\n", clock() - time);
                time = clock();
                source_down->Transform(current_transformation);
                printf("Time for transformation = %dms\n", clock() - time);
                time = clock();
                open3d::EstimateNormals(*source_down, Search_para);
                printf("Time for Normal Estimation = %dms\n", clock() - time);
                time = clock();
                open3d::RegistrationResult result_icp = open3d::RegistrationICP(
                    *source_down, *target_down, max_d,
                    Eigen::Matrix4d::Identity(),
                    open3d::TransformationEstimationPointToPlane());
                printf("Time for ICP = %dms\n", clock() - time);
                time = clock();
                current_transformation =
                    result_icp.transformation_ * current_transformation;
                mesh->Transform(result_icp.transformation_);
                printf("Time for Matrix * Matrix = %dms\n", clock() - time);
                time = clock();
                source_down->Transform(result_icp.transformation_);
                printf("Time for transformation 2 = %dms\n", clock() - time);
                time = clock();
                for (int k = 0; k < source_down->points_.size(); k++)
                    target_down->points_.push_back(source_down->points_[k]);
                printf("Time for push_back = %dms\n", clock() - time);
                time = clock();
                target_down->points_ =
                    open3d::VoxelDownSample(*target_down, vox_s)->points_;
                printf("Time for down sampling = %dms\n", clock() - time);
                time = clock();
                open3d::EstimateNormals(*target_down, Search_para);
                printf("Time for Normal Estimation = %dms\n", clock() - time);
                i++;
                printf("Time for all is %dms\n", clock() - timetime);
                return true;
            } else {
                return false;
            }
        },
        "Result");

    return 1;
}
